### PR TITLE
fix(cover,recommendation): stop metadata errors from hiding covers, improve book recommendation score logic, and prevent sitemap SQL injection

### PR DESCRIPTION
**Pull request by WilliamAGH**
## Summary

Valid book covers were permanently hidden from search results when S3 upload metadata persistence encountered transient errors, because those errors were incorrectly classified as recordable download failures. Recommendation scores from books matching multiple strategies (author + category + text) were silently dropped instead of accumulated. The sitemap query builder accepted unsanitized caller-supplied strings directly in SQL. This PR fixes all three, centralizes frontend cover URL resolution into a single priority chain, decomposes RecommendationService from ~700 to ~350 lines, and adds oxlint with Zod-specific rules to the frontend.

### Bug Fixes
- **Recommendation scores silently dropped during multi-strategy merge**: `ScoredBook.mergeWith()` return value was discarded by the `Collectors.toMap` merge function, so books matching multiple strategies always kept only the first score instead of accumulating (`RecommendationService.collectScoredCandidates`, `ScoredBook::mergeWith`)
- **Transient metadata errors permanently hid valid covers from search**: metadata persistence failures and non-persistable upload details threw `IllegalStateException`, which mapped to `UNEXPECTED_FAILURE` via the default case in `resolveUploadFailureDetail`; since `UNEXPECTED_FAILURE` was in the recordable allowlist, `recordDownloadError` was called, and `BookQueryCoverNormalizer` filters on `download_error IS NULL` — permanently excluding valid already-uploaded covers. Fixed by restricting the allowlist to 4 genuine image-failure codes: `DOWNLOAD_FAILED`, `PROCESSING_FAILED`, `TOO_LARGE`, `UNSAFE_URL` (`CoverS3UploadCoordinator.resolveRecordableErrorMessage`)
- **S3 upload failures silently swallowed**: non-persistable upload results incremented a failure counter but never propagated the error, so the reactive pipeline completed "successfully" with no cover and no error visible. Restructured to throw on validation/persistence failures so `onErrorResume` can record the error (`CoverS3UploadCoordinator.executeUpload`)
- **Successful S3 upserts did not clear stale download errors**: added `download_error = NULL` to `upsertImageLink` ON CONFLICT clause so re-uploads clear previous failures (`CoverPersistenceService.upsertImageLink`)
- **Null cause message crashed upload failure detail resolution**: `getCause().getMessage()` could NPE when the cause had no message (`CoverS3UploadCoordinator.resolveUploadFailureDetail`)
- **Cover path comparison was case-sensitive**: `needsFallback` compared URLs against the cover path segment case-sensitively, missing uppercase variants (`BookQueryCoverNormalizer`)
- **Frontend cover URL validation rejected valid timezone offsets**: `generatedAt` timestamps with `+00:00` offsets were rejected by the Zod schema (`schemas.ts`)
- **`recordDownloadError` silently returned on invalid input**: null bookId or blank error string caused a silent return with no diagnostic trace (`CoverPersistenceService.recordDownloadError`)
- **Dead null checks on DI-provided fields masked configuration errors**: 5 null guards on injected services could never trigger but silently returned empty results instead of failing fast (`RecommendationService`)

### Security
- **SQL injection in sitemap query builder**: `SitemapBookLastModifiedSqlSupport` formatted caller-supplied strings directly into SQL via `String.formatted()` without validation — now validates against `SQL_IDENTIFIER_PATTERN` and `SQL_PLACEHOLDER_PATTERN` before formatting (`SitemapBookLastModifiedSqlSupport`)

### Refactoring
- **Recommendation service decomposed from ~700 to ~350 lines**: scoring algorithms, search orchestration, and `ScoredBook` extracted into `RecommendationScoringStrategy`, `RecommendationStrategyChain`, and a top-level immutable record — each independently testable and under the LOC1 ceiling
- **Cover display URL resolved in one place instead of three**: duplicated priority chains across `BookSimilarBooks.svelte`, `searchPageViewModel.ts`, and `searchHitNormalization.ts` replaced by a computed `displayUrl` field on `CoverSchema` via Zod transform, with `buildCover()` factory for manual construction sites (`schemas.ts`)
- **S3 error codes converted from 15 string constants to typed enum**: compile-time grouping and exhaustiveness guarantee for the error taxonomy (`CoverS3UploadCoordinator.S3UploadErrorCode`)
- **Cover persistence separates validation from side effects**: `validateGoogleImageEntry` extracted as pure validation returning `Optional`, upsert call moved to caller (`CoverPersistenceService`)

### Tooling
- **Frontend linting with oxlint + eslint-plugin-zod**: 5 error-level rules catch logic bugs (z.any(), redundant .optional().default(), legacy z.number().int()), 7 warn-level rules enforce conventions; wired into `make lint`, `npm validate`, and prek pre-push hooks (`config/oxlintrc.json`, `prek.toml`). (Includes `eslint` as a devDependency at `^9.39.2` for plugin compatibility.)

### Tests
- **ScoredBook**: score accumulation, reason merging/deduplication, immutability, defensive copy (`ScoredBookTest`)
- **RecommendationScoringStrategy**: category overlap, keyword extraction, scoring constants (`RecommendationScoringStrategyTest`)
- **resolveCoverDisplayUrl**: priority chain, CoverSchema transform, buildCover factory (`schemas.test.ts`)
- **SQL injection prevention**: injection attempts, blank inputs, valid identifiers (`SitemapBookLastModifiedSqlSupportTest`)
- **S3 upload error recording**: positive test proving recordable errors (UnsafeUrlException) trigger `recordDownloadError`, negative tests confirming metadata failures do not (`CoverS3UploadCoordinatorTest`)

## Breaking Changes
None

---

> [!NOTE]
> **Medium Risk**
> Touches cover ingestion/persistence and sitemap SQL generation, so mistakes could affect cover visibility or sitemap output; changes are mostly defensive/validation-oriented but span both backend pipelines and frontend schema normalization.
> 
> **Overview**
> **Cover pipeline reliability:** Refactors `CoverS3UploadCoordinator` to use typed error codes and to fail fast on non-persistable upload results/metadata persistence issues, while only recording *true image failures* (download/processing/too-large/unsafe URL) so transient persistence problems no longer mark covers as permanently bad.
> 
> **Frontend cover normalization + tooling:** Introduces a computed `cover.displayUrl` (Zod transform + `buildCover`) and updates consumers/tests to use it, fixing edge cases like blank `preferredUrl`; adds `oxlint` (with Zod rules) to `make lint`, `npm validate`, and pre-push hooks, and bumps `zod` to v4.
> 
> **Data selection & security hardening:** Tightens AI SEO backfill eligibility to only standalone or primary-edition books and makes the backfill runner treat “description too short” as a skip; adds validation in `SitemapBookLastModifiedSqlSupport` to prevent SQL injection via formatted identifiers/placeholders; improves cover fallback URL filtering/normalization (configurable path segment + case-insensitive checks).
> 
> <sup>Written by [Cursor Bugbot](https://cursor.com/dashboard?tab=bugbot) for commit fb764e04a266911fa51ac4b8425b0da208d7cc32. This will update automatically on new commits. Configure [here](https://cursor.com/dashboard?tab=bugbot).</sup>

### CODE REVIEW

File: frontend/src/lib/components/BookSimilarBooks.svelte [hunk 1]
```diff
@@ -71,7 +60,7 @@
    * Combines: renderable URL + non-placeholder + valid aspect ratio + no runtime failure.
    */
   function hasDisplayableCover(book: Book): boolean {
-    const url = activeCoverUrlsByBookId.get(book.id) ?? resolveCoverUrl(book);
+    const url = activeCoverUrlsByBookId.get(book.id) ?? book.cover?.preferredUrl ?? null;
```
[the remark this answers is not in the document]
Comment (WilliamAGH) [reply frontend/src/lib/components/BookSimilarBooks.svelte]: # Suggestion Implemented ✅

## Changes Made
- Same fix as the CodeRabbit thread — replaced all 4 occurrences of `preferredUrl` with `displayUrl` in `BookSimilarBooks.svelte`

## Validation
- `svelte-check` passed with 0 errors, 0 warnings

File: src/main/java/net/findmybook/service/RecommendationService.java [hunk 2]
```diff
@@ -1,8 +1,12 @@
...
- * Service for generating book recommendations based on various similarity criteria
+ * Service for generating book recommendations based on various similarity criteria.
  * This class is responsible for generating book recommendations using a combination of strategies:
- * 
+ *
  * - Uses multi-faceted approach combining author, category, and text-based matching
+ *
+ * @implNote LOC1 split plan (701 lines): extract {@code RecommendationScoringStrategy}
+ *     (scoring, ranking, and deduplication logic) and {@code BookSearchStrategyChain}
+ *     (author/category/text search orchestration).
```
[the remark this answers is not in the document]
Comment (WilliamAGH) [reply src/main/java/net/findmybook/service/RecommendationService.java]: # Suggestion Implemented ✅

## Changes Made
- Replaced the stale `@implNote` referencing the extraction plan with a concise Javadoc describing the current composition: `RecommendationScoringStrategy` + `RecommendationStrategyChain`

## Validation
- `./gradlew clean test` passed (691 tests, 0 failures)

File: src/main/java/net/findmybook/support/sitemap/SitemapBookLastModifiedSqlSupport.java [hunk 3]
```diff
@@ -234,19 +225,54 @@ private static String unionAllChangeEvents() {
...
+            throw new IllegalArgumentException(
+                parameterName + " must be a valid SQL identifier matching " + SQL_IDENTIFIER_REGEX);
+        }
+    }
+
+    private static void validateSqlPlaceholders(String value) {
+        if (value == null || value.isBlank() || !SQL_PLACEHOLDER_PATTERN.matcher(value).matches()) {
+            throw new IllegalArgumentException(
+                "authorPlaceholders must contain only '?', ',', and whitespace");
+        }
```
[the remark this answers is not in the document]
Comment (WilliamAGH) [reply src/main/java/net/findmybook/support/sitemap/SitemapBookLastModifiedSqlSupport.java]: # Suggestion Deferred 📋
Valid tightening. The current regex is permissive but the inputs come from internal callers passing JDBC-style placeholders. Deferring to a separate PR to avoid broadening this PR's scope, and to pair with updated tests for the stricter validation.

File: src/main/java/net/findmybook/service/RecommendationService.java [hunk 4]
```diff
@@ -598,10 +411,7 @@ private Mono<List<Book>> searchBooks(String query, String langCode, int limit) {
             })
             .defaultIfEmpty(Collections.emptyList())
             .map(books -> limitResults(books, safeLimit))
-            .onErrorMap(error -> new IllegalStateException(
-                "RecommendationService.searchBooks query=" + query + " lang=" + langCode,
-                error
-            ));
+            .onErrorMap(error -> logAndWrapError(error, "RecommendationService.searchBooks query=" + query + " lang=" + langCode));
```
[the remark this answers is not in the document]
Comment (WilliamAGH) [reply src/main/java/net/findmybook/service/RecommendationService.java]: # Suggestion Implemented ✅

## Changes Made
The dead methods (`searchBooks`, `limitResults`, `orderBooksBySearchResults`) were already removed from RecommendationService in a prior commit. The file is now 363 lines, well within the LOC1 ceiling.

## Validation
- `./gradlew clean test` passed (691 tests, 0 failures)

File: src/main/java/net/findmybook/application/cover/CoverS3UploadCoordinator.java [hunk 5]
```diff
@@ -246,109 +269,98 @@ private void handleUploadSuccess(ImageDetails details, String bookId, UUID bookU
...
-            case CoverDownloadException downloadException -> {
-                String causeMessage = downloadException.getCause() != null
-                    ? downloadException.getCause().getMessage()
-                    : "Unknown cause";
-                yield "DownloadFailed: " + causeMessage;
+            case CoverDownloadException e -> {
+                String cause = e.getCause() != null ? e.getCause().getMessage() : "Unknown cause";
+                yield new UploadFailureDetail(S3UploadErrorCode.DOWNLOAD_FAILED,
+                    cause + " imageUrl=" + e.getImageUrl(), false);
             }
```
[the remark this answers is not in the document]
Comment (WilliamAGH) [reply src/main/java/net/findmybook/application/cover/CoverS3UploadCoordinator.java]: # Suggestion Implemented ✅

## Changes Made
- Added null guard for `e.getCause().getMessage()` — now checks both `e.getCause() != null` and `e.getCause().getMessage() != null` before using the message, falling back to `"Unknown cause"` when either is null.

## Validation
- `./gradlew clean test` — 691 tests pass, 0 failures

File: src/main/java/net/findmybook/repository/BookQueryCoverNormalizer.java [hunk 6]
```diff
@@ -38,7 +38,13 @@ final class BookQueryCoverNormalizer {
...
     private static final String LOOPBACK_IP_PREFIX = "://127.0.0.1";
     private static final String ANY_LOCAL_IP_PREFIX = "://0.0.0.0";
-    private static final String COVER_PATH_SEGMENT = "/images/book-covers/";
+    private static final String COVER_PATH_SEGMENT = System.getProperty("app.cover.path.segment", "/images/book-covers/");
+
+    /** SQL LIKE patterns with wildcards pre-baked for direct use in queries. */
+    private static final String PLACEHOLDER_COVER_LIKE = "%" + PLACEHOLDER_COVER_PATTERN + "%";
+    private static final String PRINTSEC_TITLEPAGE_LIKE = "%printsec=titlepage%";
+    private static final String PRINTSEC_COPYRIGHT_LIKE = "%printsec=copyright%";
+    private static final String PRINTSEC_TOC_LIKE = "%printsec=toc%";
```
[the remark this answers is not in the document]
Comment (WilliamAGH) [reply src/main/java/net/findmybook/repository/BookQueryCoverNormalizer.java]: # Suggestion Implemented ✅

## Changes Made
- Applied `.toLowerCase()` on `COVER_PATH_SEGMENT` in the `needsFallback()` comparison, so custom system property values with mixed casing are matched correctly against the lowercased URL.

## Validation
- `./gradlew clean test` — 691 tests pass, 0 failures

File: frontend/src/lib/validation/schemas.ts [hunk 7]
```diff
@@ -196,12 +238,12 @@ export const SitemapAuthorSchema = z.object({
...
 export const SitemapPayloadSchema = z.object({
   viewType: z.enum(["authors", "books"]),
-  activeLetter: z.string(),
   pageNumber: z.number(),
   totalPages: z.number(),
   totalItems: z.number(),
   letters: z.array(z.string()),
-  baseUrl: z.string(),
+  activeLetter: z.string(),
+  baseUrl: z.url(),
```
[the remark this answers is not in the document]
Comment (WilliamAGH) [reply frontend/src/lib/validation/schemas.ts]: # Suggestion Analysis 🔍

## Decision: Not Implemented

## Reasoning
Verified empirically that `z.url()` in Zod v4 (zod@^3.25.76, imported via `"zod/v4"`) is a **string** schema type — it accepts and validates plain strings, not URL instances:

```js
const { z } = require('zod/v4');
z.url().parse('https://example.com'); // ✅ returns string
```

The `z.url()` schema type resolves to `string` internally. It is functionally equivalent to `z.string().url()` in this Zod version. JSON string payloads from the backend parse correctly.

## Conclusion
Current implementation with `z.url()` is correct and idiomatic Zod v4. No change needed.

File: src/main/java/net/findmybook/application/cover/CoverS3UploadCoordinator.java [hunk 8]
```diff
@@ -246,109 +269,100 @@ private void handleUploadSuccess(ImageDetails details, String bookId, UUID bookU
...
+     * as {@code download_error} would hide an otherwise valid cover from queries.</p>
+     */
+    private Optional<String> resolveRecordableErrorMessage(Throwable error) {
+        UploadFailureDetail detail = resolveUploadFailureDetail(error);
+        return switch (detail.code()) {
+            case DOWNLOAD_FAILED, PROCESSING_FAILED, TOO_LARGE, UNSAFE_URL, EMPTY_UPLOAD_RESULT ->
+                Optional.of(detail.reason());
+            default -> Optional.empty();
         };
     }
```
[the remark this answers is not in the document]
Comment (WilliamAGH) [reply src/main/java/net/findmybook/application/cover/CoverS3UploadCoordinator.java]: # Suggestion Implemented ✅

## Decision
Option 2 — removed `EMPTY_UPLOAD_RESULT` (and `UNEXPECTED_FAILURE`) from the recordable allowlist.

## Reasoning
`EMPTY_UPLOAD_RESULT` is unreachable dead code in the allowlist: `emptyUploadResult()` throws plain `IllegalStateException`, which `resolveUploadFailureDetail` maps to `UNEXPECTED_FAILURE` via the `default` case. No code path ever produces an `UploadFailureDetail` with the `EMPTY_UPLOAD_RESULT` code.

`UNEXPECTED_FAILURE` was also incorrect to include — it captures metadata persistence failures and non-persistable details (both `IllegalStateException`), which would set `download_error` on the book and hide valid covers from `BookQueryCoverNormalizer`.

The recordable set is now: `DOWNLOAD_FAILED`, `PROCESSING_FAILED`, `TOO_LARGE`, `UNSAFE_URL` — genuine image failures only.

## Validation
- `./gradlew check -PskipFrontend` — 719 tests passed, 0 failures
- Added positive test verifying recordable errors DO trigger `recordDownloadError`

## Commits
- `37118d3` fix(cover): remove UNEXPECTED_FAILURE from recordable error allowlist
- `fbe753c` test(cover): add positive test for recordable error persistence
